### ISSUE TITLE
Feature Proposal: Semantic segmentation into flash cards.

### ISSUE THREAD

**Issue by Wehzie** [reorproject/reor#96]
A while ago, I had asked the following question: How can we combine the flexibility and ease of a free-form note-taking app with the learning approach of active recall and spaced repetition? A small language model driven approach comes to mind.

Currently, LogSeq offers flash cards as an integrated feature and Obsidian supports a note format inside MD files which can be extracted to Anki via a plugin, as follows.

```
TARGET DECK: {Deck name}
START
{Note Type}
{Note Fields}
Tags:
END
```

Why is this a relevant problem?
First, creating free-form notes is easier than creating Anki flash cards. I like to build a Zettelkasten; a Zettelkasten-style note-taking app is better suited for connecting information via note-links. It's also better for information lookup than Anki and other systems that use flash cards as a central data structure.
Second, I hypothesize that question variation is advantageous to reduce dependency on question cues; after extensive learning, question beginnings seem to trigger nearly automatic responses. Rephrased questions might reduce boredom when learning the same facts on short learning intervals.
Note that I have no empirical evidence for this.
Lastly, both creating good notes and good flash-cards are time intensive. Ideally, the proposed approach eliminates the time it takes to create flash cards.

How?
Freely take notes, in Reor.
Extract atomic facts; deterministic language model with temperature equal to 0.
Generate question variations for each fact with a language model.
Create a flash card for each question-fact pair.
Track knowledge progress by atomic fact, not by question.

Thoughts question difficulty.
Roar computes a similarity metric between notes.
Additionally, atomic fact similarity and distance could be computed.
Question's have different levels of difficulty.
Simple questions recall single facts.
Difficult questions span multiple facts.
Question difficulty is determined by path length between involved fact nodes.
Bridging gaps between increasingly distant facts is more difficult.

Progress visualization.
While a note-level graph would be nice, similar to Obsian or LogSeq, we could also visualize the fact graph.
We may highlight graph-subsets by learning progress.

Visual problem formulations.
A question formulation model controls a 2D canvas to create pictograms. This enables non-copyrighted visual material to be created. For example, the model could draw the leg with muscles asking for names to indexed labels.

```
ASSISTANT: What is the name of point A in the figure?
USER: It's the m. biceps brachii.
ASSISTANT: ...
```
Given correct or false responses, the user may ask follow-up questions and have a capable teacher at hand.

**Comment by samlhuillier** [comment]
This is very interesting and supporting features like this is something @chuyouchia and I have been discussing a fair bit.

If I break it down, the following is how I understand the workflow you envision:

1. Pull out atomic facts from the corpus of notes with AI
2. Generate questions for those facts also with AI
3. Compute a difficulty score based on number of facts corresponding to a question.
4. Visualise the facts in a graph perhaps to display progress.


Now to implement this there are two approaches we could take:
- Add first class support for AI flashcards and build in features that are specific to flash cards (like in logseq just with ai) .
- Build Reor in such a way that it acts as a host of general AI workflows.

If I frame the task of AI generated flash cards another way:
You start with a corpus of notes: you want to transform them into a different form (facts), then transform those facts into another form (questions) and then interact with those questions. That becomes quite a general flow that can be applied to lots of tasks, not just flash cards. For example, pulling todo items from the past month from "free-form" notes and transforming them into updated todos based on today's note or something would be a very similar task.

Reor could then become an agnostic host of AI workflows like this, providing APIs like semantic search, filesystem and database access for the LLM to call. 

The downside to this approach, particularly evident with flash cards, is how a UI would be flexible enough to support them. Though perhaps AI workflow plugins could also come with more traditional Obsidian-like plugins.

What do you think about those two approaches to this?

**Comment by Wehzie** [comment]
Thank you for kind response. I had to think a bit on this.

Let's look at the UI components that reor already supports.

1) Notes
2) Related Notes
3) Chat

Let's look at the two use cases as examples we might want to facilitate.

1) Todo aggregation
2) Question answering

Assessing subtasks and storage locations and durations:

1) Todo aggregation
	1) User invokes todo-aggregation workflow via chat.
	2) Information gathering. Read notes, notes' lifetime is user controlled.
	3) Write to current note. Current note's lifetime is user controlled.

This is easily facilitated in terms of information lifetimes, UI and storage location. If the user runs the workflow at time X, then deletes some todo in the notes and re-runs the workflow at time X+1, the aggregate todos will simply no longer show the deleted todo.

QnA is more difficult.

2) Question answering
	1) Information gathering. Reads user notes.
	2) Write atomic facts. Write to where?
	3) Question generation. Write to where?
	4) ...

Currently reor supports the note as a central data structure. Notes are created by one human user. In order to facilitate efficient llm-driven transformations of personal knowledge, the language model must gain direct write access to some kind of permanent data structure; write access must explicitly go beyond 1) copy from chat and 2) insert at cursor, both of which are mediated by a human user.

One option would be writing to the same notes the user writes to. This is disadvantageous because notes will become cluttered with abstractions, details and meta data generated by llms and thus lose their value to the human author. I propose for the llm to write to own notes which might be user editable but are clearly separated in the UI from user written notes.

2) Question answering
	1) Information gathering.
		1) Read notes and derive atomic facts, notes live as long as a user allows them to.
	2) Atomic fact writing.
		1) Write atomic facts to a "questions about X" llm-note. Inside this llm-note an atomic fact is associated with the original note by hash. The association between note and atomic fact is kept to recompute atomic facts when the underlying note changes. Furthermore, metadata about the user's memory of this atomic fact is stored alongside in the llm-note.
	3) Question generation.
		1) Cache questions by writing them to the llm-note. Tie each question to one or multiple atomic facts.
		2) Each question has a difficulty, derived from the number of question to atomic fact relations, added as metadata.
	4) User invokes question-answering workflow via chat.
		1) If no questions exist which can be retrieved from cache this triggers 1) information gathering 2) atomic fact writing and 3) question generation.
		2) Questions are selected by a scheduler. The scheduler is a llm with a prompt that is parameterized by meta data found in the llm-notes. This allows for continuity in question-to-question transition which isn't found in existing question answer systems such as Anki.
		3) The user responds via chat. A llm evaluates the response and derives a response-quality metric, for example 1 for bad response and 4 for perfect response, similar to Anki.
		4) The scheduler modifies the llm-notes to update its knowledge about the user's memory. Well remembered facts should be queried less frequently. The scheduler may draw from a config file with parameters similar to those [found in Anki](https://docs.ankiweb.net/deck-options.html#daily-limits).
				
What should a llm-note look like?
Probably JSON. Stability between user-notes and llm-notes can be achieved by letting llms write to json files with predetermined keys using [formal grammars](https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md) which are supported by llama.cpp. A reor transformation workflow then defines prompts, writeable files, and grammars which describe how files shall be written. By giving llms code access to execution their capabilites can be extended further. For example, the scheduler llm might call on a typescript scheduler script. Similarly, visualization of the current learning progress might be achieved through a script which is called by a language model, when a user asks "What is my current learning progress", opening a separate tab. So long as users can see which functions a roar workflow exposes, they have a kind of manual to go by.

When is a llm-note created?
I propose that llm-notes are workflow and workflow-argument based with their naming derived from the user command which invokes a workflow. For example, "Ask me questions about heart anatomy" creates a "anatomy questions" llm-note. If a user asks for a similar-enough workflow, existing llm-notes may be reused, else a new one is created, this makes use of semantic similarity.

In conclusion, I think that a workflow based implementation of QnA is more interesting compared to a specifically engineered solution. Further I think that the existing UI can largely accommodate QnA. Additionally necessary UI elements can be useful for other workflows.
Would other workflows benefit from such an approach? What are your thoughts on this?

**Comment by Wehzie** [comment]
Thank you for kind response. I had to think a bit on this.

Let's look at the UI components that reor already supports.

1) Notes
2) Related Notes
3) Chat

Let's look at the two use cases as examples we might want to facilitate.

1) Todo aggregation
2) Question answering

Assessing subtasks and storage locations and durations:

1) Todo aggregation
	1) User invokes todo-aggregation workflow via chat.
	2) Information gathering. Read notes, notes' lifetime is user controlled.
	3) Write to current note. Current note's lifetime is user controlled.

This is easily facilitated in terms of information lifetimes, UI and storage location. If the user runs the workflow at time X, then deletes some todo in the notes and re-runs the workflow at time X+1, the aggregate todos will simply no longer show the deleted todo.

QnA is more difficult.

2) Question answering
	1) Information gathering. Reads user notes.
	2) Write atomic facts. Write to where?
	3) Question generation. Write to where?
	4) ...

Currently reor supports the note as a central data structure. Notes are created by one human user. In order to facilitate efficient llm-driven transformations of personal knowledge, the language model must gain direct write access to some kind of permanent data structure; write access must explicitly go beyond 1) copy from chat and 2) insert at cursor, both of which are mediated by a human user.

One option would be writing to the same notes the user writes to. This is disadvantageous because notes will become cluttered with abstractions, details and meta data generated by llms and thus lose their value to the human author. I propose for the llm to write to own notes which might be user editable but are clearly separated in the UI from user written notes.

2) Question answering
	1) Information gathering.
		1) Read notes and derive atomic facts, notes live as long as a user allows them to.
	2) Atomic fact writing.
		1) Write atomic facts to a "questions about X" llm-note. Inside this llm-note an atomic fact is associated with the original note by hash. The association between note and atomic fact is kept to recompute atomic facts when the underlying note changes. Furthermore, metadata about the user's memory of this atomic fact is stored alongside in the llm-note.
	3) Question generation.
		1) Cache questions by writing them to the llm-note. Tie each question to one or multiple atomic facts.
		2) Each question has a difficulty, derived from the number of question to atomic fact relations, added as metadata.
	4) User invokes question-answering workflow via chat.
		1) If no questions exist which can be retrieved from cache this triggers 1) information gathering 2) atomic fact writing and 3) question generation.
		2) Questions are selected by a scheduler. The scheduler is a llm with a prompt that is parameterized by meta data found in the llm-notes. This allows for continuity in question-to-question transition which isn't found in existing question answer systems such as Anki.
		3) The user responds via chat. A llm evaluates the response and derives a response-quality metric, for example 1 for bad response and 4 for perfect response, similar to Anki.
		4) The scheduler modifies the llm-notes to update its knowledge about the user's memory. Well remembered facts should be queried less frequently. The scheduler may draw from a config file with parameters similar to those [found in Anki](https://docs.ankiweb.net/deck-options.html#daily-limits).
				
What should a llm-note look like?
Probably JSON. Stability between user-notes and llm-notes can be achieved by letting llms write to json files with predetermined keys using [formal grammars](https://github.com/ggerganov/llama.cpp/blob/master/grammars/README.md) which are supported by llama.cpp. A reor transformation workflow then defines prompts, writeable files, and grammars which describe how files shall be written. By giving llms code access to execution their capabilites can be extended further. For example, the scheduler llm might call on a typescript scheduler script. Similarly, visualization of the current learning progress might be achieved through a script which is called by a language model, when a user asks "What is my current learning progress", opening a separate tab. So long as users can see which functions a roar workflow exposes, they have a kind of manual to go by.

When is a llm-note created?
I propose that llm-notes are workflow and workflow-argument based with their naming derived from the user command which invokes a workflow. For example, "Ask me questions about heart anatomy" creates a "anatomy questions" llm-note. If a user asks for a similar-enough workflow, existing llm-notes may be reused, else a new one is created, this makes use of semantic similarity.

In conclusion, I think that a workflow based implementation of QnA is more interesting compared to a specifically engineered solution. Further I think that the existing UI can largely accommodate QnA. Additionally necessary UI elements can be useful for other workflows.
Would other workflows benefit from such an approach? What are your thoughts on this?

**Comment by samlhuillier** [comment]
Ok so overall if I summarise, you want to create a workflow which generates atomic notes as an intermediate artifact, then generates questions into the same LLM note for those facts. Then, you propose using leveraging AI for the question answering which gives several benefits: LLM scores a response and the LLM can produce a nice coherent ordering of questions.

To me also, it seems like generating atomic facts is almost like an intermediary/preprocessing step in an LLM pipeline to generate flashcards. You could also generate question answer pairs straightaway from notes - generating atomic facts is an improvement to help the LLM reason across multiple steps rather than just 1. I guess we'll just have to investigate which approach generates the most high quality question answer pairs. 

I like the idea of persisting the flashcards as json files. A similar approach could be used to do things like generating knowledge graphs where nodes are represented in json.
